src/datasets/utils/py_utils.py

```diff
@@ -599,6 +599,109 @@ class Pickler(dill.Pickler):
+                try:
+                    import torch
+
+                    if config.DILL_VERSION < version.parse("0.3.6"):
+
+                        @pklregister(obj_type)
+                        def _save_tensor(pickler, obj):
+                            dill._dill.log.info(f"To: {obj}")
+                            args = (obj.cpu().numpy(),)
+                            pickler.save_reduce(torch.from_numpy, args, obj=obj)
+                            dill._dill.log.info("# To")
+                            return
+
+                    elif config.DILL_VERSION.release[:3] == version.parse("0.3.6").release:
+
+                        @pklregister(obj_type)
+                        def _save_tensor(pickler, obj):
+                            dill._dill.logger.trace(pickler, f"To: {obj}")
+                            args = (obj.cpu().numpy(),)
+                            pickler.save_reduce(torch.from_numpy, args, obj=obj)
+                            dill._dill.logger.trace(pickler, "# To")
+                            return
+
+                except ImportError:
+                    pass
+            elif obj_type.__module__.startswith("spacy.lang") and any(
+                (cls.__module__, cls.__name__) == ("spacy.language", "Language") for cls in obj_type.__mro__
+            ):
+                try:
+                    import spacy
+
+                    if config.DILL_VERSION < version.parse("0.3.6"):
+
+                        @pklregister(obj_type)
+                        def _save_lang(pickler, obj):
+                            def _create_lang(config, bytes_data):
+                                lang_cls = spacy.util.get_lang_class(config["nlp"]["lang"])
+                                nlp = lang_cls.from_config(config)
+                                return nlp.from_bytes(bytes_data)
+
+                            dill._dill.log.info(f"Sp: {obj}")
+                            args = (obj.config, obj.to_bytes())
+                            pickler.save_reduce(_create_lang, args, obj=obj)
+                            dill._dill.log.info("# Sp")
+                            return
+
+                    elif config.DILL_VERSION.release[:3] == version.parse("0.3.6").release:
+
+                        @pklregister(obj_type)
+                        def _save_lang(pickler, obj):
+                            def _create_lang(config, bytes_data):
+                                lang_cls = spacy.util.get_lang_class(config["nlp"]["lang"])
+                                nlp = lang_cls.from_config(config)
+                                return nlp.from_bytes(bytes_data)
+
+                            dill._dill.logger.trace(pickler, f"Sp: {obj}")
+                            args = (obj.config, obj.to_bytes())
+                            pickler.save_reduce(_create_lang, args, obj=obj)
+                            dill._dill.logger.trace(pickler, "# Sp")
```

Comment:
it's maybe simpler to select the dill logger outside this function depending on `dill`'s version ? This way you don't have to duplicate the code.